### PR TITLE
test(docker): mocked unittests external calls 

### LINKED ISSUE #1640: move docker tests doing external calls to integration and use mocking for unit tests

### PR DESCRIPTION
**Pull request by slettner**
Patched or moved unit tests in the docker module to avoid external calls.
Some tests are now in the `integration/hub_usage/test_hub_usage.py` module.

Fixes #1640

### CODE REVIEW

**Review by JoanFM — COMMENTED** [review]
Really nice work! Thank you for the contribution!!

**Comment by JoanFM** [comment]
recheckcla

**Comment by slettner** [comment]
Thanks :)
I will have a look at the failing tests in the evening 👍🏼

**Comment by nan-wang** [comment]
**Jina CLA check**<br/>
  ❤️ Thank you for your pull request. It looks like this is your first contribution to an open source project maintained by Jina AI Limited. Before we can look at your pull request, we kindly ask that you sign our [Contributor License Agreement](https://github.com/jina-ai/jina/blob/master/.github/CLA.md). You can sign it by commenting in format below.
  - - -
  ***I have read the CLA Document and I hereby sign the CLA***
  - - -

**Comment by slettner** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by slettner** [comment]
I have seen that the CI times out after 15 minutes, I think this causes two modules to fail (`drivers` and `flow`).

For the docker module, apparently, the right images are not available leading to the `IndexError`.

```    @pytest.fixture(scope='function')
    def docker_image():
        def _filter_repo_tag(image, image_name='jinahub/pod.dummy_mwu_encoder'):
            tags = image.attrs['RepoTags'] if 'RepoTags' in image.attrs else None
            if tags:
                return tags[0].startswith(image_name)
            else:
                return False
    
        client = docker.from_env()
        images = client.images.list()
>       image_name = list(filter(lambda image: _filter_repo_tag(image), images))[0]
E       IndexError: list index out of range
```

Are these images built by other tests? Might the Error come from the execution order of tests?

**Comment by JoanFM** [comment]
> I have seen that the CI times out after 15 minutes, I think this causes two modules to fail (`drivers` and `flow`).
> 
> For the docker module, apparently, the right images are not available leading to the `IndexError`.
> 
> ```
>     def docker_image():
>         def _filter_repo_tag(image, image_name='jinahub/pod.dummy_mwu_encoder'):
>             tags = image.attrs['RepoTags'] if 'RepoTags' in image.attrs else None
>             if tags:
>                 return tags[0].startswith(image_name)
>             else:
>                 return False
>     
>         client = docker.from_env()
>         images = client.images.list()
> >       image_name = list(filter(lambda image: _filter_repo_tag(image), images))[0]
> E       IndexError: list index out of range
> ```
> 
> Are these images built by other tests? Might the Error come from the execution order of tests?

I think u may be able to find in the CI how the images are built before

**Comment by florian-hoenicke** [comment]
@slettner Can you please resolve the merge conflicts

**Comment by slettner** [comment]
Hi @florian-hoenicke,

I updated the PR after a rebase, sorry for the inactivity. 

The tests look better now but the `test_hub_build_level` fails now with an `IndexError` while fetching the local docker images. 
After debugging, I found that the images in my environment are named `jinahub/pod.encoder.dummy_mwu_encoder` instead of the current `jinahub/pod.dummy_mwu_encoder` causing the index error. 

Unfortunately, this change only fixes the first test of the module, the second one produces no fail-levels at all, where 
`expected_failed_levels = [BuildTestLevel.POD_DOCKER, BuildTestLevel.FLOW]` is expected.

Do you have an idea what might be the cause?

File: tests/integration/hub_usage/dummyhub/requirements.txt [hunk 1]
```diff
@@ -1 +1 @@
-jina
\ No newline at end of file
+jina==1.0.9
```
Comment (JoanFM) [inline tests/integration/hub_usage/dummyhub/requirements.txt:1]: can u please remove these, these changes should not be commited

File: tests/integration/hub_usage/dummyhub_pretrained/requirements.txt [hunk 2]
```diff
@@ -1 +1 @@
-jina==1.0.5
\ No newline at end of file
+jina==1.0.9
```
Comment (JoanFM) [inline tests/integration/hub_usage/dummyhub_pretrained/requirements.txt:1]: same here

File: tests/integration/hub_usage/dummyhub_slow/requirements.txt [hunk 3]
```diff
@@ -1 +1 @@
-jina
\ No newline at end of file
+jina==1.0.9
```
Comment (JoanFM) [inline tests/integration/hub_usage/dummyhub_slow/requirements.txt:1]: same here

File: tests/unit/docker/hub-mwu-bad/bad-pythonfile/Dockerfile [hunk 4]
```diff
@@ -1,5 +1,5 @@
 FROM jinaai/jina:test-pip
 
-sadasdasdADD *.py mwu_encoder.yml ./
+ADD *.py mwu_encoder.yml ./
```
Comment (JoanFM) [inline tests/unit/docker/hub-mwu-bad/bad-pythonfile/Dockerfile:3]: this is bad on purpose to fail in the test

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: tests/unit/docker/hub-mwu-bad/bad-pythonfile/Dockerfile (hunk 4 above)
Comment (slettner) [reply tests/unit/docker/hub-mwu-bad/bad-pythonfile/Dockerfile:3]: This is the directory for the `bad-pythonfile` which already contains errors.
The `Dockerfile` can be correct in that case, can it?

**Comment by florian-hoenicke** [comment]
@slettner would you like to continue working on this issue? Or should someone from us take over?

**Comment by slettner** [comment]
I can pick it up today, sorry for the inactivity :)

File: tests/unit/docker/test_hub_build.py [hunk 5]
```diff
@@ -91,6 +91,23 @@ def test_hub_build_uses():
     HubIO(args).build()
 
 
+@pytest.mark.timeout(360)
+def test_hub_build_uses():
+    args = set_hub_build_parser().parse_args(
+        [os.path.join(cur_dir, 'hub-mwu'), '--test-uses', '--raise-error']
+    )
+    assert HubIO(args).build()["is_build_success"]
```
Comment (JoanFM) [inline tests/unit/docker/test_hub_build.py:99]: please stick to single quote

File: tests/unit/docker/test_hub_list.py [hunk 6]
```diff
@@ -10,25 +11,36 @@ def test_hub_list_local_with_submodule():
     assert len(response) > 1
 
 
-def test_hub_list_keywords():
+@mock.patch('jina.docker.hubapi.remote.urlopen')
+def test_hub_list_keywords(mocker):
+    mocker.return_value.__enter__.return_value.read.return_value = json.dumps(
+        [{'name': "foo"}, {'name': "bar"}]
+    )
```
Comment (JoanFM) [inline tests/unit/docker/test_hub_list.py:18]: single quote

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: tests/unit/docker/test_hub_build.py (hunk 5 above)
Comment (slettner) [reply tests/unit/docker/test_hub_build.py]: https://github.com/jina-ai/jina/pull/1958/commits/7301267d0a05735aa9a251c3f25dab8b1bb3d9be

File: tests/unit/docker/test_hub_list.py (hunk 6 above)
Comment (slettner) [reply tests/unit/docker/test_hub_list.py:18]: https://github.com/jina-ai/jina/pull/1958/commits/7301267d0a05735aa9a251c3f25dab8b1bb3d9be

File: tests/unit/docker/hub-mwu-bad/bad-pythonfile/Dockerfile (hunk 4 above)
Comment (slettner) [reply tests/unit/docker/hub-mwu-bad/bad-pythonfile/Dockerfile]: I reverted it in https://github.com/jina-ai/jina/pull/1958/commits/a47b970e35bec04f58f8b38de5f8c0c9324b6ef7

**Review by florian-hoenicke — APPROVED** [review]
Good work!

**Review by JoanFM — APPROVED** [review]

**Pull request merged**